File: apps/api/src/routers/webhooks.py
"""Webhook subscription and delivery management endpoints.

Dashboard-only (JWT). Tenants subscribe to events for their own tenant_id.
"""

import logging

from fastapi import APIRouter, Depends, HTTPException, Query, status

from src.core.dependencies import AgentDependencies
from src.core.deps import get_deps
from src.core.tenant import get_tenant_id_from_jwt
from src.models.api import MessageResponse
from src.models.webhook import (
    WEBHOOK_EVENTS,
    CreateWebhookRequest,
    CreateWebhookResponse,
    TestFireRequest,
    UpdateWebhookRequest,
    WebhookDeliveryListResponse,
    WebhookDeliveryResponse,
    WebhookListResponse,
    WebhookResponse,
)
from src.services.webhook import (
    WebhookLimitExceeded,
    WebhookService,
    WebhookURLInvalid,
)
from src.services.webhook_delivery import WebhookDeliveryService

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v1/webhooks", tags=["webhooks"])


def _get_webhook_service(deps: AgentDependencies = Depends(get_deps)) -> WebhookService:
    return WebhookService(webhooks_collection=deps.webhooks_collection)


def _get_delivery_service(
    deps: AgentDependencies = Depends(get_deps),
) -> WebhookDeliveryService:
    return WebhookDeliveryService(deliveries_collection=deps.webhook_deliveries_collection)


@router.get("/events", response_model=list[str])
async def list_event_types(
    _: str = Depends(get_tenant_id_from_jwt),
) -> list[str]:
    """Return the available event types tenants can subscribe to."""
    return list(WEBHOOK_EVENTS)


@router.post("", response_model=CreateWebhookResponse, status_code=201)
async def create_webhook(
    body: CreateWebhookRequest,
    tenant_id: str = Depends(get_tenant_id_from_jwt),
    service: WebhookService = Depends(_get_webhook_service),
) -> CreateWebhookResponse:
    try:
        response, secret = await service.create(tenant_id=tenant_id, body=body)
    except WebhookLimitExceeded as e:
        raise HTTPException(status_code=409, detail=str(e))
    except WebhookURLInvalid as e:
        raise HTTPException(status_code=422, detail=str(e))
    return CreateWebhookResponse(secret=secret, **response)


@router.get("", response_model=WebhookListResponse)
async def list_webhooks(
    tenant_id: str = Depends(get_tenant_id_from_jwt),
    service: WebhookService = Depends(_get_webhook_service),
) -> WebhookListResponse:
    items = await service.list_for_tenant(tenant_id)
    return WebhookListResponse(webhooks=[WebhookResponse(**i) for i in items])


@router.get("/{webhook_id}", response_model=WebhookResponse)
async def get_webhook(
    webhook_id: str,
    tenant_id: str = Depends(get_tenant_id_from_jwt),
    service: WebhookService = Depends(_get_webhook_service),
) -> WebhookResponse:
    doc = await service.get_response(webhook_id=webhook_id, tenant_id=tenant_id)
    if not doc:
        raise HTTPException(status_code=404, detail="Webhook not found")
    return WebhookResponse(**doc)


@router.patch("/{webhook_id}", response_model=WebhookResponse)
async def update_webhook(
    webhook_id: str,
    body: UpdateWebhookRequest,
    tenant_id: str = Depends(get_tenant_id_from_jwt),
    service: WebhookService = Depends(_get_webhook_service),
) -> WebhookResponse:
    try:
        result = await service.update(webhook_id=webhook_id, tenant_id=tenant_id, body=body)
    except WebhookURLInvalid as e:
        raise HTTPException(status_code=422, detail=str(e))
    if result is None:
        raise HTTPException(status_code=404, detail="Webhook not found")
    return WebhookResponse(**result)


@router.delete("/{webhook_id}", response_model=MessageResponse)
async def delete_webhook(
    webhook_id: str,
    tenant_id: str = Depends(get_tenant_id_from_jwt),
    service: WebhookService = Depends(_get_webhook_service),
) -> MessageResponse:
    deleted = await service.delete(webhook_id=webhook_id, tenant_id=tenant_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Webhook not found")
    return MessageResponse(message="Webhook deleted")


@router.post("/{webhook_id}/test", response_model=MessageResponse, status_code=202)
async def test_fire_webhook(
    webhook_id: str,
    body: TestFireRequest,
    tenant_id: str = Depends(get_tenant_id_from_jwt),
    service: WebhookService = Depends(_get_webhook_service),
    delivery: WebhookDeliveryService = Depends(_get_delivery_service),
    deps: AgentDependencies = Depends(get_deps),
) -> MessageResponse:
    """Fire a synthetic event payload to a single subscribed webhook."""
    webhook = await service.get(webhook_id=webhook_id, tenant_id=tenant_id)
    if not webhook:
        raise HTTPException(status_code=404, detail="Webhook not found")
    if body.event not in webhook["events"]:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Webhook is not subscribed to this event",
        )
    sample_data = {"test": True, "message": "This is a sample webhook delivery"}
    await delivery.deliver(webhook=webhook, event=body.event, data=sample_data)
    return MessageResponse(message="Test event delivered")


@router.get(
    "/{webhook_id}/deliveries",
    response_model=WebhookDeliveryListResponse,
)
async def list_webhook_deliveries(
    webhook_id: str,
    limit: int = Query(default=50, ge=1, le=200),
    tenant_id: str = Depends(get_tenant_id_from_jwt),
    service: WebhookService = Depends(_get_webhook_service),
    delivery: WebhookDeliveryService = Depends(_get_delivery_service),
) -> WebhookDeliveryListResponse:
    # Existence + tenant ownership check.
    webhook = await service.get(webhook_id=webhook_id, tenant_id=tenant_id)
    if not webhook:
        raise HTTPException(status_code=404, detail="Webhook not found")
    items = await delivery.list_recent(tenant_id=tenant_id, webhook_id=webhook_id, limit=limit)
    return WebhookDeliveryListResponse(deliveries=[WebhookDeliveryResponse(**i) for i in items])

File: apps/web/lib/webhooks.ts
/**
 * Typed client functions for the FastAPI /api/v1/webhooks endpoints.
 *
 * Server-only — these helpers mint a backend JWT, so they must never run in
 * the browser. Use them from Server Components or Server Actions.
 */

import "server-only";

import { apiFetch } from "@/lib/api-client";

export const WEBHOOK_EVENTS = [
  "document.ingested",
  "document.deleted",
  "chat.completed",
  "subscription.updated",
] as const;

export type WebhookEvent = (typeof WEBHOOK_EVENTS)[number];

export interface Webhook {
  id: string;
  url: string;
  events: WebhookEvent[];
  description: string | null;
  active: boolean;
  secret_prefix: string;
  created_at: string;
  updated_at: string;
}

export interface CreatedWebhook extends Webhook {
  secret: string;
}

export interface WebhookDelivery {
  id: string;
  webhook_id: string;
  event: string;
  status: "pending" | "delivered" | "failed";
  attempts: number;
  response_code: number | null;
  last_error: string | null;
  created_at: string;
  delivered_at: string | null;
}

export interface CreateWebhookInput {
  url: string;
  events: WebhookEvent[];
  description?: string;
  active?: boolean;
}

export interface UpdateWebhookInput {
  url?: string;
  events?: WebhookEvent[];
  description?: string | null;
  active?: boolean;
}

export async function listWebhooks(): Promise<Webhook[]> {
  const data = await apiFetch<{ webhooks: Webhook[] }>("/api/v1/webhooks");
  return data.webhooks;
}

export async function createWebhook(
  input: CreateWebhookInput,
): Promise<CreatedWebhook> {
  return apiFetch<CreatedWebhook>("/api/v1/webhooks", {
    method: "POST",
    body: input,
  });
}

export async function updateWebhook(
  webhookId: string,
  input: UpdateWebhookInput,
): Promise<Webhook> {
  return apiFetch<Webhook>(`/api/v1/webhooks/${webhookId}`, {
    method: "PATCH",
    body: input,
  });
}

export async function deleteWebhook(webhookId: string): Promise<void> {
  await apiFetch<{ message: string }>(`/api/v1/webhooks/${webhookId}`, {
    method: "DELETE",
  });
}

export async function testFireWebhook(
  webhookId: string,
  event: WebhookEvent,
): Promise<void> {
  await apiFetch<{ message: string }>(
    `/api/v1/webhooks/${webhookId}/test`,
    { method: "POST", body: { event } },
  );
}

export async function listWebhookDeliveries(
  webhookId: string,
): Promise<WebhookDelivery[]> {
  const data = await apiFetch<{ deliveries: WebhookDelivery[] }>(
    `/api/v1/webhooks/${webhookId}/deliveries?limit=25`,
  );
  return data.deliveries;
}
